services/web/server/src/simcore_service_webserver/garbage_collector_tasks_api_keys.py

```diff
@@ -0,0 +1,72 @@
+"""
+    Scheduled task that periodically runs prune in the garbage collector service
+
+"""
+import asyncio
+import logging
+from typing import AsyncIterator, Callable
+
+from aiohttp import web
+from tenacity import retry
+from tenacity.before_sleep import before_sleep_log
+from tenacity.wait import wait_exponential
+
+from .login.api_keys_db import prune_expired_api_keys
+
+logger = logging.getLogger(__name__)
+
+CleanupContextFunc = Callable[[web.Application], AsyncIterator[None]]
+
+_SEC = 1
+
+_PERIODIC_TASK_NAME = f"{__name__}.prune_expired_api_keys_periodically"
+_APP_TASK_KEY = f"{_PERIODIC_TASK_NAME}.task"
+
+
+@retry(
+    wait=wait_exponential(min=5 * _SEC),
+    before_sleep=before_sleep_log(logger, logging.WARNING),
+)
+async def _run_periodically(app: web.Application, wait_period_s: float):
+    """Periodically check expiration dates and updates user status
+
+    It is resilient, i.e. if update goes wrong, it waits a bit and retries
+    """
+
+    while True:
+        if deleted := await prune_expired_api_keys(app):
+
+            # broadcast force logout of user_id
+            for api_key in deleted:
+                logger.info("API-key %s expired and was removed", f"{api_key=}")
+
+        else:
+            logger.info("No API keys expired")
+
+        await asyncio.sleep(wait_period_s)
```

Comment:
We already have a discussion about this pattern [here](https://github.com/ITISFoundation/osparc-simcore/pull/3380#discussion_r979593787) so let's finish it up there.

Reply:
updated